design/keycloak.x/operator.md

```diff
@@ -0,0 +1,128 @@
+* The Keycloak community is more keen to Java and there is less Go Lang expertise
+* The current project needs some high-cost maintenance tasks to be performed in order to use the latest features (e.g. webhooks) and receive the latest fixes and patches, specifically:
+    * upgrading Go lang version from 1.13 to 1.17
+    * upgrading the Operator SDK and the dependencies
+
+  Those upgrades will require creating a completely new project, using different libraries, moving, and in some cases, rewriting the components, e.g. the whole testsuite.
+* The current approach around CRDs no longer fits the long term vision for cloud-native Keycloak as it is very error-prone and fragile.
+* A Java operator can share business objects with the Keycloak main codebase increasing the code-reuse and dramatically reducing the chances of introducing bugs in the translation process from Kubernetes resources.
+* A unified codebase will make it easy to test and debug the entire system.
+* A new operator will embrace from the ground up the new Cloud Native capabilities of upcoming Keycloak releases such as the Quarkus distribution and Store.X, making those first-class citizen overall improving the user experience.
+
+
+## Features
+
+---
+**NOTE**
+
+The primary target of the operator is to make it easy to achieve production grade installations of Keycloak.X.
+
+---
+
+### Configuring Keycloak deployment
+
+The operator will use a CRD representing the Keycloak installation. The CRD will expose the following configuration options:
+* custom Keycloak image; default base image will be used if not specified
+* Git configuration to fetch static business objects configuration (see below)
+* manual horizontal and vertical scaling
+* pod and node affinity, toleration
+* SSL config, truststore
+
+Since most of the configuration options will come from the Keycloak.X distribution itself, the CRD will also expose appropriate fields for passing any distribution related options to the container, like database connection details (obviously without any credentials), SPIs configuration, etc.
+
+### Configuring Keycloak business objects using Kubernetes resources
+
+The new operator will expose two main ways of configuring business objects in Keycloak (e.g.: Realms, Roles, Clients, etc.) in addition to the built-in Dynamic configuration through the REST API/UI console:
+* Static file based configuration stored in Git to enable GitOps operations
+* Static configuration through Kubernetes API
+
+Static configuration will be strictly read-only, therefore two-way syncing is not going to be needed.
+
+Static configuration is going to provide an immutable and cloud native way for managing business objects in Keycloak that can be easily moved between environments (e.g. dev, stage, prod) in a predictable manner. This feature will leverage the new Store.X which enables federation of the configuration from multiple sources (static/dynamic) by re-architecting the storage layer.
```

Comment:
Can you add a link to the appropriate Store.X capabilities here? I think this is rather vague and led me to "I want to know more".

Reply:
@hmlnarik @mhajas What is the best "summary" doc on Store.X? :) There are several design proposals (some even unmerged yet) but is there anything that would give the whole picture?